writer.go

```diff
@@ -637,15 +637,21 @@ func (w *Writer) WriteMessages(ctx context.Context, msgs ...Message) error {
 	// to increasing GC work.
 	assignments := make(map[topicPartition][]int32)
 
+	topicPartitionCount := make(map[string]int)
+
 	for i, msg := range msgs {
 		topic, err := w.chooseTopic(msg)
 		if err != nil {
 			return err
 		}
 
-		numPartitions, err := w.partitions(ctx, topic)
-		if err != nil {
-			return err
+		numPartitions, exists := topicPartitionCount[topic]
+		if !exists {
```

Comment:
you can combine above two lines - if <line 648>; !exists { ... }

Reply:
when i do that, Go doesn't like the numPartitions i have outside the if statement (at 657)